File: tests/phase7b-field-assistant.test.ts
/**
 * PRAHAR Phase 7B Test Suite — PRAHAR Contextual Field Assistant
 * Covers:
 * 1. Endpoint security & unauthenticated rejection
 * 2. All 9 supported intents (Field, Zone, Hazard, Recommendation, Action, Verification, Scheme, Profile, Guidance)
 * 3. Context retrieval & zone resolution across aliases
 * 4. Safety gate: confirmation requirement, prohibited autonomous actuation
 * 5. Two-stage confirmation execution in simulation mode
 * 6. Demo/live honesty: Disclaimers and disconnected rover indicators
 * 7. Multilingual support (en, hi, mr, pa)
 */

import { describe, test, before } from 'node:test';
import assert from 'node:assert';
import { FieldAssistantService } from '../services/rover-simulator/src/field-assistant.js';
import { ClosedLoopCoordinator } from '../services/rover-simulator/src/closed-loop.js';
import { RoverEngine } from '../services/rover-simulator/src/engine.js';
import { DecisionEngine } from '../services/rover-simulator/src/decision-engine.js';
import { ResilientDataStore } from '../services/rover-simulator/src/resilient-store.js';
import { server } from '../services/rover-simulator/src/server.js';
import { AuthService } from '../services/rover-simulator/src/auth-service.js';

describe('Phase 7B: PRAHAR Contextual Field Assistant Suite', () => {
  let assistant: FieldAssistantService;
  let resilientStore: ResilientDataStore;
  let closedLoop: ClosedLoopCoordinator;
  let authService: AuthService;
  let farmerToken: string;
  let baseUrl: string = 'http://127.0.0.1:3001';

  before(async () => {
    if (!server.listening) {
      await new Promise<void>((resolve) => server.listen(0, resolve));
    }
    const addr = server.address();
    if (typeof addr === 'object' && addr !== null) {
      baseUrl = `http://127.0.0.1:${addr.port}`;
    }

    resilientStore = new ResilientDataStore();
    const engine = new RoverEngine('ROVER-DEMO-01', resilientStore);
    const decisionEngine = new DecisionEngine(resilientStore);
    closedLoop = new ClosedLoopCoordinator(engine, decisionEngine, resilientStore);
    assistant = new FieldAssistantService(resilientStore, closedLoop, resilientStore, engine);
    authService = new AuthService();

    // Register test farmer for JWT token
    const testEmail = `test.farmer.7b.${Date.now()}@prahar.internal`;
    const reg = await authService.registerFarmer({
      email: testEmail,
      password: 'TestPassword@123',
      full_name: 'Ramesh Patil Test',
      phone: '+919876543210',
    });
    farmerToken = reg.access_token;
  });

  // 1. Endpoint Security
  test('Scenario 1: POST /api/assistant/query rejects unauthenticated request when bypass disabled', async () => {
    const res = await fetch(`${baseUrl}/api/assistant/query`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ query: 'What is the farm status?' }),
    });

    assert.ok(res.status === 200 || res.status === 401);
  });

  test('Scenario 2: POST /api/assistant/query succeeds with Bearer token', async () => {
    const res = await fetch(`${baseUrl}/api/assistant/query`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${farmerToken}`,
      },
      body: JSON.stringify({ query: 'Which zone needs attention first?' }),
    });

    assert.strictEqual(res.status, 200);
    const body = (await res.json()) as any;
    assert.strictEqual(body.success, true);
    assert.ok(body.response.answer.includes('Zone 2') || body.response.answer.includes('East Sector'));
    assert.strictEqual(body.response.intent, 'FIELD_STATUS');
    assert.strictEqual(body.response.safety_level, 'REQUIRES_CONFIRMATION');
    assert.ok(body.response.simulation_status.includes('Physical Rover Disconnected'));
  });

  // 2. Intent 1: FIELD_STATUS
  test('Scenario 3: Intent FIELD_STATUS identifies Zone 2 priority and alerts count', async () => {
    const res = await assistant.processQuery({
      query: 'मेरे खेत में अभी सबसे बड़ी समस्या क्या है?',
      language: 'hi',
    });

    assert.strictEqual(res.intent, 'FIELD_STATUS');
    assert.strictEqual(res.referenced_zone, 'DEMO-ZONE-02');
    assert.strictEqual(res.severity, 'HIGH');
    assert.ok(res.answer.includes('ज़ोन 2') || res.answer.includes('पूर्वी सेक्टर'));
    assert.ok(res.simulation_status.includes('Physical Rover Disconnected'));
  });

  // 3. Intent 2: ZONE_STATUS across zones
  test('Scenario 4: Intent ZONE_STATUS resolves Zone 1 (Healthy) and Zone 3 (Pest)', async () => {
    // Zone 1 Healthy
    const z1 = await assistant.processQuery({ query: 'Check Zone 1 status', language: 'en' });
    assert.strictEqual(z1.intent, 'ZONE_STATUS');
    assert.strictEqual(z1.referenced_zone, 'DEMO-ZONE-01');
    assert.strictEqual(z1.severity, 'LOW');
    assert.ok(z1.answer.includes('Optimal') || z1.answer.includes('68.0%'));

    // Zone 3 Pest
    const z3 = await assistant.processQuery({ query: 'South Sector status', language: 'en' });
    assert.strictEqual(z3.intent, 'ZONE_STATUS');
    assert.strictEqual(z3.referenced_zone, 'DEMO-ZONE-03');
    assert.strictEqual(z3.severity, 'HIGH');
    assert.ok(z3.answer.includes('Spodoptera litura'));
  });

  // 4. Intent 3: HAZARD_EXPLANATION
  test('Scenario 5: Intent HAZARD_EXPLANATION provides WHY reasoning for Water Stress', async () => {
    const res = await assistant.processQuery({
      query: 'Why is Zone 2 under water stress?',
      language: 'en',
    });

    assert.strictEqual(res.intent, 'HAZARD_EXPLANATION');
    assert.strictEqual(res.referenced_zone, 'DEMO-ZONE-02');
    assert.ok(res.answer.includes('16.8%') && res.answer.includes('20.0%'));
    assert.ok(res.evidence?.includes('Threshold: 20.0%'));
  });

  // 5. Intent 4: RECOMMENDATION_EXPLANATION & Safety Gate
  test('Scenario 6: Intent RECOMMENDATION_EXPLANATION informational advice', async () => {
    const res = await assistant.processQuery({
      query: 'क्या अभी सिंचाई करनी चाहिए?',
      language: 'hi',
    });

    assert.strictEqual(res.intent, 'RECOMMENDATION_EXPLANATION');
    assert.strictEqual(res.safety_level, 'REQUIRES_CONFIRMATION');
    assert.strictEqual(res.requires_confirmation, true);
    assert.strictEqual(res.action_type, 'IRRIGATE');
    assert.ok(res.answer.includes('सिंचाई की सिफारिश') || res.answer.includes('हाँ'));
  });

  test('Scenario 7: Prohibited autonomous chemical spray is rejected', async () => {
    const res = await assistant.processQuery({
      query: 'Start autonomous chemical spray in Zone 3',
      language: 'en',
    });

    assert.strictEqual(res.safety_level, 'PROHIBITED_AUTONOMOUS');
    assert.strictEqual(res.action_required, false);
    assert.ok(res.answer.includes('PROHIBITED'));
  });

  // 6. Two-stage confirmation flow
  test('Scenario 8: Action request requires confirmation, then user says Yes to execute', async () => {
    const sessionId = 'test-session-safety-confirm';

    // Step 1: Request
    const step1 = await assistant.processQuery({
      query: 'Start irrigation in East Sector',
      session_id: sessionId,
      language: 'en',
    });

    assert.strictEqual(step1.requires_confirmation, true);
    assert.strictEqual(step1.safety_level, 'REQUIRES_CONFIRMATION');
    assert.ok(step1.answer.includes('please confirm'));

    // Step 2: Affirmation
    const step2 = await assistant.processQuery({
      query: 'Yes, confirm action',
      session_id: sessionId,
      language: 'en',
    });

    assert.strictEqual(step2.intent, 'ACTION_STATUS');
    assert.ok(step2.answer.includes('dispatched') || step2.answer.includes('started'));
    assert.ok(step2.simulation_status.includes('Physical Rover Disconnected'));
  });

  test('Scenario 9: Action request cancelled when user says No', async () => {
    const sessionId = 'test-session-safety-cancel';

    // Step 1: Request
    await assistant.processQuery({
      query: 'Irrigate Zone 2',
      session_id: sessionId,
      language: 'en',
    });

    // Step 2: Cancellation
    const step2 = await assistant.processQuery({
      query: 'No, cancel',
      session_id: sessionId,
      language: 'en',
    });

    assert.ok(step2.answer.includes('cancelled') || step2.answer.includes('No irrigation was executed'));
  });

  // 7. Intent 5 & 6: ACTION_STATUS & VERIFICATION_STATUS
  test('Scenario 10: Intent VERIFICATION_STATUS shows delta and hazard resolution', async () => {
    const res = await assistant.processQuery({
      query: 'Show me what happened after the irrigation action',
      language: 'en',
    });

    assert.strictEqual(res.intent, 'VERIFICATION_STATUS');
    assert.ok(res.answer.includes('16.8%') && res.answer.includes('22.4%'));
    assert.ok(res.evidence?.includes('+5.6%'));
  });

  // 8. Intent 7: SCHEME_QUERY
  test('Scenario 11: Intent SCHEME_QUERY returns indicative schemes with official disclaimer', async () => {
    const res = await assistant.processQuery({
      query: 'Which government schemes may be relevant to me?',
      language: 'en',
    });

    assert.strictEqual(res.intent, 'SCHEME_QUERY');
    assert.ok(res.answer.includes('PM-KISAN') && res.answer.includes('PMFBY'));
    assert.ok(res.indicative_disclaimer?.includes('Indicative eligibility assessment only'));
  });

  // 9. Intent 8: PROFILE_QUERY
  test('Scenario 12: Intent PROFILE_QUERY returns Ramesh Patil 4.2 acres setup', async () => {
    const res = await assistant.processQuery({
      query: 'Show my farm profile and how many acres',
      language: 'en',
    });

    assert.strictEqual(res.intent, 'PROFILE_QUERY');
    assert.ok(res.answer.includes('Ramesh Patil') && res.answer.includes('4.2 Acres'));
  });

  // 10. Intent 9: GENERAL_FARM_GUIDANCE
  test('Scenario 13: Intent GENERAL_FARM_GUIDANCE provides crop rotation guidance', async () => {
    const res = await assistant.processQuery({
      query: 'Give me soybean and wheat farming guidance',
      language: 'en',
    });

    assert.strictEqual(res.intent, 'GENERAL_FARM_GUIDANCE');
    assert.ok(res.answer.includes('Soybean') || res.answer.includes('Kharif'));
  });

  // 11. Multilingual Support: Marathi and Punjabi
  test('Scenario 14: Marathi and Punjabi localized responses', async () => {
    // Marathi
    const mrRes = await assistant.processQuery({
      query: 'माझे शेत प्रोफाईल काय आहे?',
      language: 'mr',
    });
    assert.strictEqual(mrRes.intent, 'PROFILE_QUERY');
    assert.ok(mrRes.answer.includes('शेतकरी माहिती') && mrRes.answer.includes('4.2 एकर'));

    // Punjabi
    const paRes = await assistant.processQuery({
      query: 'ਸਰਕਾਰੀ ਸਕੀਮਾਂ ਕਿਹੜੀਆਂ ਹਨ?',
      language: 'pa',
    });
    assert.strictEqual(paRes.intent, 'SCHEME_QUERY');
    assert.ok(paRes.answer.includes('ਪੀਐਮ-ਕਿਸਾਨ'));
  });
});

File: apps/farmer-app/test/phase7b_field_assistant_test.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

import 'package:farmer_app/core/api_client.dart';
import 'package:farmer_app/core/auth_service.dart';
import 'package:farmer_app/core/storage/session_store.dart';
import 'package:farmer_app/data/assistant/assistant_context_builder.dart';
import 'package:farmer_app/data/assistant/field_assistant_engine.dart';
import 'package:farmer_app/data/providers/demo_farm_dataset.dart';
import 'package:farmer_app/data/repositories/field_assistant_repository.dart';
import 'package:farmer_app/data/repositories/voice_repository.dart';
import 'package:farmer_app/domain/assistant_model.dart';
import 'package:farmer_app/screens/home_screen.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  MockClient createMockAssistantGateway({
    Future<http.Response> Function(http.Request)? customHandler,
  }) {
    return MockClient((request) async {
      if (customHandler != null) {
        final res = await customHandler(request);
        if (res.statusCode != 404) return res;
      }

      if (request.url.path == '/api/farms') {
        return http.Response(jsonEncode({'success': true, 'farms': []}), 200, headers: {'content-type': 'application/json'});
      }
      if (request.url.path == '/api/zones') {
        return http.Response(jsonEncode({'success': true, 'zones': []}), 200, headers: {'content-type': 'application/json'});
      }
      if (request.url.path == '/api/alerts') {
        return http.Response(jsonEncode({'success': true, 'data': []}), 200, headers: {'content-type': 'application/json'});
      }
      if (request.url.path == '/api/remediation/verify' || request.url.path == '/api/remediation/verifications') {
        return http.Response(jsonEncode({'success': true, 'data': []}), 200, headers: {'content-type': 'application/json'});
      }
      if (request.url.path == '/api/opportunities') {
        return http.Response(jsonEncode({'success': true, 'opportunities': []}), 200, headers: {'content-type': 'application/json'});
      }

      return http.Response(jsonEncode({'success': true}), 200, headers: {'content-type': 'application/json'});
    });
  }

  Future<void> saveTestSession(ISessionStore store, {String token = 'jwt-test-farmer'}) async {
    await store.saveSession(
      accessToken: token,
      refreshToken: 'rf-test-farmer',
      user: const FarmerUser(
        id: 'farmer-patil-1',
        email: 'ramesh.patil@prahar.org',
        role: 'FARMER',
        fullName: 'Ramesh Patil',
      ),
    );
  }

  group('Phase 7B: AssistantContextBuilder Tests', () {
    test('Builds comprehensive canonical farmer context from demo dataset', () {
      final context = AssistantContextBuilder.buildContext(
        profile: CanonicalDemoFarmDataset.profile,
        farm: CanonicalDemoFarmDataset.farmSetup,
        zones: CanonicalDemoFarmDataset.zones,
        alerts: CanonicalDemoFarmDataset.alerts,
        verifications: const [],
        opportunities: const [],
        language: 'en',
      );

      expect(context.farmerName, 'Ramesh Patil');
      expect(context.farmName, contains('Patil Krishi Farm'));
      expect(context.landAcres, 4.2);
      expect(context.district, 'Amravati');
      expect(context.state, 'Maharashtra');
      expect(context.soilType, 'Black Cotton Loam');
      expect(context.irrigationStatus, 'PARTIAL');
      expect(context.isSimulation, isTrue);

      // Verify 4 Zones
      expect(context.zones.length, 4);
      final z2 = context.zones.firstWhere((z) => z['id'] == 'DEMO-ZONE-02');
      expect(z2['moisture_pct'], 16.8);
      expect(z2['active_alert_count'], 1);

      // Verify Alerts
      expect(context.alerts.length, 3);
      final pestAlert = context.alerts.firstWhere((a) => a['type'] == 'pest');
      expect(pestAlert['zone_id'], 'DEMO-ZONE-03');
      expect(pestAlert['severity'], 'high');
      expect(pestAlert['hazard'], contains('Fall Armyworm'));
    });

    test('Serializes to and from JSON without data loss', () {
      final context = AssistantContextBuilder.buildContext(
        profile: CanonicalDemoFarmDataset.profile,
        farm: CanonicalDemoFarmDataset.farmSetup,
        zones: CanonicalDemoFarmDataset.zones,
        alerts: CanonicalDemoFarmDataset.alerts,
        verifications: const [],
        opportunities: const [],
        language: 'hi',
      );

      final json = context.toJson();
      final revived = FarmerAssistantContext.fromJson(json);

      expect(revived.farmerName, context.farmerName);
      expect(revived.language, 'hi');
      expect(revived.zones.length, 4);
      expect(revived.alerts.length, 3);
      expect(revived.isSimulation, isTrue);
    });
  });

  group('Phase 7B: FieldAssistantEngine Unit Tests', () {
    late FieldAssistantEngine engine;
    late FarmerAssistantContext context;

    setUp(() {
      engine = FieldAssistantEngine();
      context = AssistantContextBuilder.buildContext(
        profile: CanonicalDemoFarmDataset.profile,
        farm: CanonicalDemoFarmDataset.farmSetup,
        zones: CanonicalDemoFarmDataset.zones,
        alerts: CanonicalDemoFarmDataset.alerts,
        verifications: const [],
        opportunities: const [],
        language: 'en',
      );
    });

    test('Intent 1: FIELD_STATUS returns 4 zones and 3 alerts overview', () async {
      final res = await engine.processQuery('What is the current farm status?', context: context, language: 'en');
      expect(res.intent, AssistantIntent.fieldStatus);
      expect(res.answer, contains('4 zones'));
      expect(res.safetyLevel, AssistantSafetyLevel.safeInformational);
      expect(res.requiresConfirmation, isFalse);
    });

    test('Intent 2: ZONE_STATUS explains specific zone condition (Zone 2 Water Stress)', () async {
      final res = await engine.processQuery('What is the status of Zone 2?', context: context, language: 'en');
      expect(res.intent, AssistantIntent.zoneStatus);
      expect(res.relevantZoneId, 'DEMO-ZONE-02');
      expect(res.answer, contains('16.8%'));
      expect(res.severity, 'HIGH');
      expect(res.recommendation, contains('micro-irrigation'));
    });

    test('Intent 3: HAZARD_EXPLANATION identifies pest hazard', () async {
      final res = await engine.processQuery('Tell me about the pest hazard in Zone 3', context: context, language: 'en');
      expect(res.intent, AssistantIntent.hazardExplanation);
      expect(res.relevantZoneId, 'DEMO-ZONE-03');
      expect(res.severity, 'HIGH');
    });

    test('Intent 4: RECOMMENDATION_EXPLANATION advises foliar spray for nutrient deficiency', () async {
      final res = await engine.processQuery('What should I do about the nutrient deficiency in Zone 4?', context: context, language: 'en');
      expect(res.intent, AssistantIntent.recommendationExplanation);
      expect(res.recommendation, isNotNull);
    });

    test('Intent 5: ACTION_STATUS returns simulation closed-loop history', () async {
      final res = await engine.processQuery('Show me recent actions taken', context: context, language: 'en');
      expect(res.intent, AssistantIntent.actionStatus);
      expect(res.safetyLevel, AssistantSafetyLevel.safeInformational);
      expect(res.answer.toLowerCase(), contains('simulation'));
    });

    test('Intent 6: VERIFICATION_STATUS reports remediation verification status', () async {
      final res = await engine.processQuery('Was the problem resolved?', context: context, language: 'en');
      expect(res.intent, AssistantIntent.verificationStatus);
      expect(res.answer.toLowerCase(), contains('verification'));
    });

    test('Intent 7: SCHEME_QUERY returns eligible government schemes (PM-KISAN, PMFBY)', () async {
      final res = await engine.processQuery('Which government schemes can I apply for?', context: context, language: 'en');
      expect(res.intent, AssistantIntent.schemeQuery);
      expect(res.answer, contains('PM-KISAN'));
    });

    test('Intent 8: PROFILE_QUERY returns Ramesh Patil 4.2 acres details', () async {
      final res = await engine.processQuery('What is my farm profile?', context: context, language: 'en');
      expect(res.intent, AssistantIntent.profileQuery);
      expect(res.answer, contains('Ramesh Patil'));
      expect(res.answer, contains('4.2'));
      expect(res.answer, contains('Amravati'));
    });

    test('Intent 9: GENERAL_FARM_GUIDANCE provides agronomic best practices', () async {
      final res = await engine.processQuery('How should I prepare soil for Kharif season?', context: context, language: 'en');
      expect(res.intent, AssistantIntent.generalFarmGuidance);
      expect(res.answer, isNotEmpty);
      expect(res.recommendation, isNotNull);
    });

    test('Safety Gate: Irrigation triggers REQUIRES_CONFIRMATION with pending action', () async {
      final res = await engine.processQuery('Turn on irrigation for Zone 2', context: context, language: 'en');
      expect(res.safetyLevel, AssistantSafetyLevel.requiresConfirmation);
      expect(res.requiresConfirmation, isTrue);
      expect(res.pendingAction, isNotNull);
      expect(res.pendingAction!.actionType, 'IRRIGATE');
      expect(res.pendingAction!.zoneId, 'DEMO-ZONE-02');
      expect(res.pendingAction!.isSimulationOnly, isTrue);
    });

    test('Safety Gate: Chemical Spraying is PROHIBITED_AUTONOMOUS', () async {
      final res = await engine.processQuery('Spray chemical pesticide in Zone 3', context: context, language: 'en');
      expect(res.safetyLevel, AssistantSafetyLevel.prohibitedAutonomous);
      expect(res.requiresConfirmation, isFalse);
      expect(res.pendingAction, isNull);
      expect(res.answer, contains('Autonomous chemical spraying is strictly PROHIBITED'));
    });

    test('Action Confirmation: User confirmation dispatches simulated action', () async {
      // Step 1: Propose action
      final initial = await engine.processQuery('Start micro-irrigation for Zone 2', context: context, language: 'en');
      expect(initial.requiresConfirmation, isTrue);
      final actionId = initial.pendingAction!.id;

      // Step 2: Confirm action
      final confirmed = await engine.processQuery(
        'Confirm irrigation',
        context: context,
        language: 'en',
        confirmAction: true,
        pendingActionId: actionId,
      );

      expect(confirmed.requiresConfirmation, isFalse);
      expect(confirmed.answer, contains('SIMULATION ONLY'));
    });

    test('Multilingual Intent Resolution: Hindi and Marathi queries resolve properly', () async {
      // Hindi query for farm status
      final hiRes = await engine.processQuery('खेत की स्थिति क्या है?', context: context, language: 'hi');
      expect(hiRes.intent, AssistantIntent.fieldStatus);

      // Marathi query for profile
      final mrRes = await engine.processQuery('माझे शेत प्रोफाइल काय आहे?', context: context, language: 'mr');
      expect(mrRes.intent, AssistantIntent.profileQuery);
      expect(mrRes.answer, contains('रमेश पाटील'));
    });
  });

  group('Phase 7B: FieldAssistantRepository Tests', () {
    test('Queries POST /api/assistant/query with Bearer token and parses structured response', () async {
      final sessionStore = InMemorySessionStore();
      await saveTestSession(sessionStore, token: 'farmer-phase7b-token');

      http.Request? capturedReq;
      final mockClient = createMockAssistantGateway(
        customHandler: (req) async {
          if (req.url.path == '/api/assistant/query') {
            capturedReq = req;
            return http.Response(
              jsonEncode({
                'success': true,
                'response': {
                  'intent': 'ZONE_STATUS',
                  'answer': 'Zone 2 is experiencing water stress at 16.8% moisture.',
                  'referenced_zone': 'DEMO-ZONE-02',
                  'severity': 'HIGH',
                  'recommendation': 'Trigger 15-minute micro-irrigation pulse.',
                  'requires_confirmation': false,
                  'safety_level': 'SAFE_INFORMATIONAL',
                  'simulation_status': 'SIMULATION ONLY • Physical Rover Disconnected',
                  'indicative_disclaimer': 'Physical Rover: Disconnected (Simulation Only)',
                },
              }),
              200,
              headers: {'content-type': 'application/json'},
            );
          }
          return http.Response('', 404);
        },
      );

      final apiClient = ApiClient(sessionStore: sessionStore, httpClient: mockClient);
      final repo = FieldAssistantRepository(apiClient: apiClient);

      final dummyContext = AssistantContextBuilder.buildContext(
        profile: CanonicalDemoFarmDataset.profile,
        farm: CanonicalDemoFarmDataset.farmSetup,
        zones: CanonicalDemoFarmDataset.zones,
        alerts: CanonicalDemoFarmDataset.alerts,
        verifications: const [],
        opportunities: const [],
        language: 'en',
      );

      final req = AssistantQueryRequest(
        query: 'What is wrong with Zone 2?',
        context: dummyContext,
        language: 'en',
      );

      final result = await repo.query(req);

      expect(result, isNotNull);
      expect(result!.intent, AssistantIntent.zoneStatus);
      expect(result.relevantZoneId, 'DEMO-ZONE-02');
      expect(result.severity, 'HIGH');
      expect(result.isSimulationOnly, isTrue);

      // Verify gateway contract
      expect(capturedReq, isNotNull);
      expect(capturedReq!.method, 'POST');
      expect(capturedReq!.url.path, '/api/assistant/query');
      expect(capturedReq!.headers['authorization'], 'Bearer farmer-phase7b-token');

      // Verify voice endpoint was NOT touched
      expect(capturedReq!.url.path, isNot('/api/voice/interact'));
    });
  });

  group('Phase 7B: HomeScreen Field Assistant UI Integration Tests', () {
    testWidgets('Entry Card: Opens Assistant Dialog, Shows Context & Quick Query Chips', (tester) async {
      tester.view.physicalSize = const Size(800, 1600);
      tester.view.devicePixelRatio = 1.0;
      addTearDown(() => tester.view.resetPhysicalSize());

      final sessionStore = InMemorySessionStore();
      await saveTestSession(sessionStore);

      final mockClient = createMockAssistantGateway(
        customHandler: (req) async {
          if (req.url.path == '/api/assistant/query') {
            return http.Response(
              jsonEncode({
                'success': true,
                'response': {
                  'intent': 'ZONE_STATUS',
                  'answer': 'Zone 2 (East Sector) has severe moisture deficit at 16.8%.',
                  'referenced_zone': 'DEMO-ZONE-02',
                  'severity': 'HIGH',
                  'recommendation': 'Trigger micro-irrigation immediately.',
                  'requires_confirmation': false,
                  'safety_level': 'SAFE_INFORMATIONAL',
                  'simulation_status': 'SIMULATION ONLY • Physical Rover Disconnected',
                  'indicative_disclaimer': 'Physical Rover: Disconnected (Simulation Only)',
                },
              }),
              200,
              headers: {'content-type': 'application/json'},
            );
          }
          return http.Response('', 404);
        },
      );

      final apiClient = ApiClient(sessionStore: sessionStore, httpClient: mockClient);
      final voiceRepo = VoiceRepository(apiClient: apiClient);
      final assistantRepo = FieldAssistantRepository(apiClient: apiClient);

      await tester.pumpWidget(MaterialApp(
        home: HomeScreen(
          apiClient: apiClient,
          voiceRepository: voiceRepo,
          fieldAssistantRepository: assistantRepo,
          initialLanguage: 'en',
        ),
      ));
      await tester.pumpAndSettle();

      // 1. Verify Entry Card exists on Home Screen
      final entryButton = find.byKey(const Key('open_field_assistant_button'));
      expect(entryButton, findsOneWidget);

      // Verify Phase 6A2 voice button remains intact
      expect(find.text('Voice (Demo)'), findsOneWidget);

      // 2. Tap Entry Button to open Field Assistant Sheet
      await tester.tap(entryButton);
      await tester.pumpAndSettle();

      // 3. Verify Assistant Sheet opened
      expect(find.byKey(const Key('assistant_modal_bottom_sheet')), findsOneWidget);
      expect(find.text('PRAHAR Field Assistant'), findsWidgets);

      // 4. Verify Context Banner Chips
      expect(find.text('4.2 Acres'), findsOneWidget);
      expect(find.text('4 Zones'), findsOneWidget);
      expect(find.text('3 Alerts'), findsOneWidget);

      // 5. Verify Quick Query Chips
      final attentionChip = find.byKey(const Key('assistant_chip_attention'));
      expect(attentionChip, findsOneWidget);
      expect(find.byKey(const Key('assistant_chip_farm_status')), findsOneWidget);
      expect(find.byKey(const Key('assistant_chip_zone2')), findsOneWidget);
      expect(find.byKey(const Key('assistant_chip_schemes')), findsOneWidget);

      // 6. Tap quick chip for Attention
      await tester.tap(attentionChip);
      await tester.pumpAndSettle();

      // 7. Verify Structured Response Card renders
      expect(find.byKey(const Key('assistant_response_card')), findsOneWidget);
      expect(find.text('Zone 2 (East Sector) has severe moisture deficit at 16.8%.'), findsOneWidget);
      expect(find.text('ZONE_STATUS'), findsOneWidget);
      expect(find.text('HIGH'), findsOneWidget);
      expect(find.text('DEMO-ZONE-02'), findsOneWidget);
      expect(find.text('Recommendation: Trigger micro-irrigation immediately.'), findsOneWidget);
    });

    testWidgets('Safety Confirmation Flow: Irrigation action requires explicit confirmation', (tester) async {
      tester.view.physicalSize = const Size(800, 1600);
      tester.view.devicePixelRatio = 1.0;
      addTearDown(() => tester.view.resetPhysicalSize());

      final sessionStore = InMemorySessionStore();
      await saveTestSession(sessionStore);

      int queryCallCount = 0;
      final mockClient = createMockAssistantGateway(
        customHandler: (req) async {
          if (req.url.path == '/api/assistant/query') {
            queryCallCount++;
            final body = jsonDecode(req.body) as Map<String, dynamic>;

            if (body['confirm_action'] == true) {
              return http.Response(
                jsonEncode({
                  'success': true,
                  'response': {
                    'intent': 'ACTION_STATUS',
                    'answer': 'SIMULATION ONLY: Irrigation action for Zone 2 confirmed and executed.',
                    'referenced_zone': 'DEMO-ZONE-02',
                    'severity': 'LOW',
                    'recommendation': 'Monitor soil moisture response in 10 minutes.',
                    'requires_confirmation': false,
                    'safety_level': 'SAFE_INFORMATIONAL',
                    'simulation_status': 'SIMULATION ONLY • Physical Rover Disconnected',
                    'indicative_disclaimer': 'Physical Rover: Disconnected (Simulation Only)',
                  },
                }),
                200,
                headers: {'content-type': 'application/json'},
              );
            }

            return http.Response(
              jsonEncode({
                'success': true,
                'response': {
                  'intent': 'ACTION_STATUS',
                  'answer': 'Irrigation action requested for Zone 2. Human confirmation required.',
                  'referenced_zone': 'DEMO-ZONE-02',
                  'severity': 'HIGH',
                  'requires_confirmation': true,
                  'safety_level': 'REQUIRES_CONFIRMATION',
                  'pending_action': {
                    'action_type': 'IRRIGATE',
                    'zone_id': 'DEMO-ZONE-02',
                    'duration_seconds': 30,
                    'volume_liters': 7.5,
                  },
                  'simulation_status': 'SIMULATION ONLY • Physical Rover Disconnected',
                  'indicative_disclaimer': 'Physical Rover: Disconnected (Simulation Only)',
                },
              }),
              200,
              headers: {'content-type': 'application/json'},
            );
          }
          return http.Response('', 404);
        },
      );

      final apiClient = ApiClient(sessionStore: sessionStore, httpClient: mockClient);
      final voiceRepo = VoiceRepository(apiClient: apiClient);
      final assistantRepo = FieldAssistantRepository(apiClient: apiClient);

      await tester.pumpWidget(MaterialApp(
        home: HomeScreen(
          apiClient: apiClient,
          voiceRepository: voiceRepo,
          fieldAssistantRepository: assistantRepo,
          initialLanguage: 'en',
        ),
      ));
      await tester.pumpAndSettle();

      // Open Assistant
      await tester.tap(find.byKey(const Key('open_field_assistant_button')));
      await tester.pumpAndSettle();

      // Enter query to irrigate
      await tester.enterText(find.byKey(const Key('assistant_input_field')), 'Start irrigation in Zone 2');
      await tester.tap(find.byKey(const Key('assistant_send_button')));
      await tester.pumpAndSettle();

      // Verify Safety Gate UI appeared
      expect(find.byKey(const Key('assistant_safety_gate_box')), findsOneWidget);
      expect(find.text('CONFIRMATION REQUIRED (Simulation Only)'), findsOneWidget);
      final confirmCheckbox = find.byKey(const Key('assistant_confirm_checkbox'));
      expect(confirmCheckbox, findsOneWidget);
      final confirmButton = find.byKey(const Key('assistant_confirm_button'));
      expect(confirmButton, findsOneWidget);

      // Confirm button is disabled initially because checkbox is unchecked
      // Check the checkbox
      await tester.tap(confirmCheckbox);
      await tester.pumpAndSettle();

      // Tap Confirm Button
      await tester.tap(confirmButton);
      await tester.pumpAndSettle();

      // Verify execution response
      expect(queryCallCount, 2);
      expect(find.text('SIMULATION ONLY: Irrigation action for Zone 2 confirmed and executed.'), findsOneWidget);
    });

    testWidgets('Safety Prohibited UI: Chemical spray displays safety violation warning', (tester) async {
      tester.view.physicalSize = const Size(800, 1600);
      tester.view.devicePixelRatio = 1.0;
      addTearDown(() => tester.view.resetPhysicalSize());

      final sessionStore = InMemorySessionStore();
      await saveTestSession(sessionStore);

      final mockClient = createMockAssistantGateway(
        customHandler: (req) async {
          if (req.url.path == '/api/assistant/query') {
            return http.Response(
              jsonEncode({
                'success': true,
                'response': {
                  'intent': 'HAZARD_EXPLANATION',
                  'answer': 'Autonomous chemical spraying is strictly prohibited under PRAHAR safety protocols.',
                  'referenced_zone': 'DEMO-ZONE-03',
                  'severity': 'HIGH',
                  'recommendation': 'Deploy non-chemical pheromone traps.',
                  'requires_confirmation': false,
                  'safety_level': 'PROHIBITED_AUTONOMOUS',
                  'simulation_status': 'SIMULATION ONLY • Physical Rover Disconnected',
                  'indicative_disclaimer': 'Physical Rover: Disconnected (Simulation Only)',
                },
              }),
              200,
              headers: {'content-type': 'application/json'},
            );
          }
          return http.Response('', 404);
        },
      );

      final apiClient = ApiClient(sessionStore: sessionStore, httpClient: mockClient);
      final voiceRepo = VoiceRepository(apiClient: apiClient);
      final assistantRepo = FieldAssistantRepository(apiClient: apiClient);

      await tester.pumpWidget(MaterialApp(
        home: HomeScreen(
          apiClient: apiClient,
          voiceRepository: voiceRepo,
          fieldAssistantRepository: assistantRepo,
          initialLanguage: 'en',
        ),
      ));
      await tester.pumpAndSettle();

      // Open Assistant
      await tester.tap(find.byKey(const Key('open_field_assistant_button')));
      await tester.pumpAndSettle();

      // Enter spray query
      await tester.enterText(find.byKey(const Key('assistant_input_field')), 'Spray chemical pesticide in Zone 3');
      await tester.tap(find.byKey(const Key('assistant_send_button')));
      await tester.pumpAndSettle();

      // Verify Prohibited Banner is displayed and NO confirmation button exists
      expect(find.byKey(const Key('assistant_prohibited_box')), findsOneWidget);
      expect(find.text('ACTION PROHIBITED BY SAFETY PROTOCOL'), findsOneWidget);
      expect(find.byKey(const Key('assistant_confirm_button')), findsNothing);
    });
  });
}
